src/mdl/holder.rs

```diff
@@ -0,0 +1,533 @@
+    let mdoc = crate::credential::mdoc::Mdoc::from_cbor_encoded_document(mdoc_bytes, key_alias)
+        .map_err(|e| SessionError::Generic {
+            value: format!("Error retrieving MDoc from storage: {e:}"),
+        })?;
+    let drms = DeviceRetrievalMethods::new(DeviceRetrievalMethod::BLE(BleOptions {
+        peripheral_server_mode: None,
+        central_client_mode: Some(CentralClientMode { uuid }),
+    }));
+    let session = SessionManagerInit::initialise(
+        NonEmptyMap::new("org.iso.18013.5.1.mDL".into(), mdoc.document().clone()),
+        Some(drms),
+        None,
+    )
+    .map_err(|e| SessionError::Generic {
+        value: format!("Could not initialize session: {e:?}"),
+    })?;
+    let ble_ident = session
+        .ble_ident()
+        .map_err(|e| SessionError::Generic {
+            value: format!("Couldn't get BLE identification: {e:?}").to_string(),
+        })?
+        .to_vec();
+    let (engaged_state, qr_code_uri) =
+        session.qr_engagement().map_err(|e| SessionError::Generic {
+            value: format!("Could not generate qr engagement: {e:?}"),
+        })?;
+    Ok(MdlPresentationSession {
+        engaged: Mutex::new(Some(engaged_state)),
+        in_process: Mutex::new(None),
+        qr_code_uri,
+        ble_ident,
+    })
+}
+
+#[derive(uniffi::Object)]
+pub struct MdlPresentationSession {
+    engaged: Mutex<Option<device::SessionManagerEngaged>>,
+    in_process: Mutex<Option<InProcessRecord>>,
+    pub qr_code_uri: String,
+    pub ble_ident: Vec<u8>,
+}
+
+#[derive(uniffi::Object, Clone)]
+struct InProcessRecord {
+    session: device::SessionManager,
+    items_request: device::RequestedItems,
+}
+
+#[uniffi::export]
+impl MdlPresentationSession {
+    /// Handle a request from a reader that is seeking information from the mDL holder.
+    ///
+    /// Takes the raw bytes received from the reader by the holder over the transmission
+    /// technology. Returns a Vector of information items requested by the reader, or an
+    /// error.
+    pub fn handle_request(&self, request: Vec<u8>) -> Result<Vec<ItemsRequest>, RequestError> {
+        let (session_manager, items_requests) = {
+            let session_establishment: SessionEstablishment = serde_cbor::from_slice(&request)
+                .map_err(|e| RequestError::Generic {
+                    value: format!("Could not deserialize request: {e:?}"),
+                })?;
+            // Mutexes only return Err if another thread has panicked while holding the mutex
+            // If that has happened, its probably better to just crash. This is what the standard documentation recommends.
+            // See https://doc.rust-lang.org/std/sync/struct.Mutex.html
+            let engaged = self
+                .engaged
+                .lock()
+                .expect("Could not lock mutex")
+                .deref_mut()
+                .take();
+            engaged
+                .unwrap()
```

Comment:
If you can remove the `Option` from `self.engaged` then this is a moot point, otherwise could you add a comment of the form `// Unwrap safety: <reason>` on the line directly above the unwrap call?